### ISSUE TITLE
Allow annotation of any UniProtKB identifier

### ISSUE THREAD

**Issue by cmungall** [geneontology/noctua#471]
This could be one of a number of different ways

 1. [ ] allow pasting in of a UniProtKB ID in all the various relevant forms (annoton wizards, add individual, TableMode, complex editor), regardless of whether it is in neo
 2. [ ] load all of uniprot into neo
 3. [ ] change architecture such that external autocomplete services can be used an client can inject TBox axioms (minimally rdfs:label)

It would seem 1 is easiest, we should target this first. I believe we discussed this at the geneva noctua workshop and implemented it, but it appears to have gone.

to be discussed further today @kltm @vanaukenk @ukemi @pgaudet

**Comment by kltm** [comment]
I'd disagree--the last two seem easier. Number two would work within what we have now, and I think that H left Minerva to take external services. Number one would require a widget-level coordination, contraction, and refactoring that would take quite a bit of effort.

**Comment by cmungall** [comment]
To clarify, 1 and 2 are standalone solutions. solution 1 in isolation requires curators to memorize IDs both for entry and for recognizing IDs in the display.

solution 3 depends on solution 1. The curator still has to inject an ID, but once injected it will be visible via labels in the normal way

**Comment by cmungall** [comment]
Additional clarification: here any uniprotkb ID means both:

 * so-called canonical IDs, e.g. P12345
 * isoform IDs, e.g. P12345-2

**Comment by cmungall** [comment]
if we went route3 and plugged in external APIs for either AC or lookup mygene is the cleanest, but may not be complete: https://github.com/biothings/mygene.info/issues/19

**Comment by kltm** [comment]
Your characterization of 1 is correct from a curator point of view. 1 is conceptually standalone, but would require either a fair amount of client code coordination or the creation of an uber autocomplete (on the wishlist) which then needs to be uniformly rolled out. I believe that we want to get this solved before that would likely be completed.
3 still requires 1's "any id" policy, but users would still possibly have the labels after minerva did a lookup, so that is an improvement.
2 is the cleanest in that we keep everything in-house, which also makes things like label updates, a possibility in 3, pretty much a non-issue.

I'm unsure about the use of mygene out of the box for 3. For example, if I put in "fox2" (http://mygene.info/v3/query?q=fox2), I get a single "close" result for human, whereas amigo gives 3 exact and one partial results and neo gives an SGD exact and a few partials.

**Comment by cmungall** [comment]
> I'm unsure about the use of mygene out of the box
good point, it would be better for lookup given an ID rather than for AC

**Comment by cmungall** [comment]
Note to self: behavior with unknown IDs can be tested with M3ExpressionParserTest

**Comment by kltm** [comment]
https://github.com/geneontology/noctua/issues/471#issuecomment-324489696
Does that include round-tripping, or just initial ingest?

**Comment by kltm** [comment]
These issues seem to be the last ones we were looking at in this neighborhood:
https://github.com/geneontology/minerva/issues/53
https://github.com/geneontology/minerva/issues/58
We currently run with `run-minerva-no-validation`, which is: lookup `yes` and validation: `no`.

**Comment by kltm** [comment]
As it currently stands, assuming you have entered *something* in the GP (enabled_by) entry in the "Add annoton" section, free entry/copying and pasting is allowed for all entries.
Everything that is not in the "Add annoton" section is restricted to a selection.

**Comment by cmungall** [comment]
To summarize:

Number 1 is 'good enough' for a first pass for SIB curators. It turns out that this is possible with the add-annoton wizard. I personally don't use this much, and I've observed a variety of behaviors from different users. I filed #473 which I think will make add-annoton more generally usable.

For the next pass, we should think about labels, but we have some breathing space now (would be happy to close this one and start a new ticket)

**Comment by kltm** [comment]
For number 1, any further action will take some time or hinge on #261 .

**Comment by thomaspd** [comment]
I'm bringing @pmasson55 (Swiss-Prot curator!) into this discussion

**Comment by kltm** [comment]
From the BioHackathon, @JervenBolleman also has an interest in this.

**Comment by pmasson55** [comment]
After few weeks of testing as curator,  the current solution of entering UniProt Accessions works as a temporary solution but it’s really limiting for the curation. For any model I create there are multiple viral proteins, and using only UniProtKB: XXXXX, the pathway becomes hard to follow and read after the addition of only few proteins. For usability we need to have protein names.
Moreover for viruses we need more than just the UniProtKB protein names.  Since viruses encode proteins as polyproteins that get cleaved post-translationally, the entities we really need to annotate are the “chains”. For a single viral entry, there can be many different chains, each representing an individual functional unit.  It would be good to be able to annotate these chains individually (by uploading them also in neo from UniProtKB ?).

**Comment by krchristie** [comment]
I completely agree that we need more than just the UniProtKB protein name, not just in the canvas but **also in the Annotation Preview**. It defeats the point of the Annotation Preview to have to export the GPAD to be certain that the correct thing has been annotated.

Also, it would be nice if the autocomplete in the _**Add Annoton**_ wizard worked the same way as in the _**Add Individual**_. Specifically _**Add Individual**_ allowed me to use "Q8N4C6-7" as the text I entered to get autocompletion upon, while entering this text into the _**Add Annoton**_ wizard only offered me ChEBI terms. In contrast, in the _**Add Annoton**_ wizard, I had to type in "Nin Hs" in order to get the needed ID offered in the autocomplete list. This text worked equally well in the _**Add Individual**_ option.

**Comment by thomaspd** [comment]
I wanted to revisit this now, so that we can use the SAE (@tmushayahama ) to annotate any UniProt entry.  I would suggest that we allow pasting of an identifier, as is done now.  But then we can use the UniProt web services to programmatically retrieve the gene name, species etc. that we can then populate onto the gene product instance.  So this information could then be viewed in Noctua.  Using web services would be straightforward for the "canonical" UniProt entry, but @pmasson55, can you suggest how to get the relevant information for "chains" from the UniProt web services?  What is an example identifier?

**Comment by pmasson55** [comment]
Hi,
Concerning chains, Poliovirus is a good example :http://www.uniprot.org/uniprot/P03300
If you go on PTM/processing part of the entry, there is a list of all the chains. For example, 
Capsid protein VP0 has a chain identifer PRO_0000424688. In ProteintoGO, we usually type the name of the entry followed by the name of the chain. So if we want to annotate this VP0 chain  specifically we would have this identifier P03300:PRO_0000424688. For the technical retrieval issue, I can ask Nicole or Jerven. 
Hope this help and answer your question.

**Comment by JervenBolleman** [comment]
Just to confirm, the input from UniProt can be an Entry (primary) accession, An isoform accession or a Chain identifier? 
The desired output are the gene names and species (NCBI tax id?) associated with this entry?

**Comment by pgaudet** [comment]
> Just to confirm, the input from UniProt can be an Entry (primary) accession, An isoform accession or a Chain identifier?
The desired output are the gene names and species (NCBI tax id?) associated with this entry?

Yes - but hopefully a single gene name and species (ie a single entity). 

Pascale

**Comment by JervenBolleman** [comment]
@pgaudet there is an 1:n relation of entry to gene name (in rare cases). 1:1 from entry to species.

**Comment by pgaudet** [comment]
But there is a primary name ? 
(in any case that would be only for display, although I am not sure how that would be managed in the back-end).

**Comment by JervenBolleman** [comment]
@pgaudet one uniprot entry can have multiple genes e.g. http://www.uniprot.org/uniprot/Q9ULZ0 (the same is true for neXtProt by the way)

See the results of this SPARQL query: for the human swiss-prot entries.
```sparql
PREFIX up:<http://purl.uniprot.org/core/> 
PREFIX taxon:<http://purl.uniprot.org/taxonomy/> 
SELECT 
	?protein 
	(COUNT(?gene) AS ?genes)
FROM <http://sparql.uniprot.org/uniprot>
WHERE	
{
	?protein a up:Protein ;
               up:organism taxon:9606 ;
               up:reviewed true ;
               up:encodedBy ?gene 
} 
GROUP BY ?protein
HAVING (COUNT(?gene) > 1)
```

**Comment by cmungall** [comment]
Is there a way to query for whether a protein is a GCRP? These should generally have one gene

**Comment by pmasson55** [comment]
Hello,

It may happen but it's very rare. I usually try to gather annotations in the best representative genomes that are well annotated. So, I would say it represents less than 1% of the annotations  in conventional GO.

**Comment by vanaukenk** [comment]
I'd like to add this to the agenda for our GO-CAM call on Wednesday, April 25th, just to make sure we're all clear on the allowed and expected behaviors for annotating to UniProtKB accessions.  Thx.

**Comment by pgaudet** [comment]
@alexsign will look into providing the GPI for all the Swiss-Prot entries.

**Comment by kltm** [comment]
@pgaudet Is this now functionally a dupe of https://github.com/geneontology/neo/issues/82?

**Comment by pgaudet** [comment]
Replaced by https://github.com/geneontology/neo/issues/82
